### PR TITLE
Allow actors to target terrain without force-fire

### PR DESCRIPTION
**Pull request by obrakmann**
This is a small addition to `AttackBase`. It should be useful for things like mortars, artillery or ballistic missiles.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Attack/AttackBase.cs [hunk 1]
```diff
@@ -55,6 +55,10 @@ public abstract class AttackBaseInfo : PausableConditionalTraitInfo
 		[Desc("Tolerance for attack angle. Range [0, 512], 512 covers 360 degrees.")]
 		public readonly WAngle FacingTolerance = new(512);
 
+		[Desc("Show the target cursor even on empty terrain cells when enabled.",
+			"Any traits implementing IMove must be absent or disabled for this to work.")]
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Traits/Attack/AttackBase.cs]: But it works just fine? You can still order around units with force move in necessary

File: OpenRA.Mods.Common/Traits/Attack/AttackBase.cs [hunk 2]
```diff
@@ -477,8 +481,8 @@ bool CanTargetLocation(Actor self, CPos location, TargetModifiers modifiers, ref
 
 				IsQueued = modifiers.HasModifier(TargetModifiers.ForceQueue);
 
-				// Targeting the terrain is only possible with force-attack modifier
-				if (modifiers.HasModifier(TargetModifiers.ForceMove) || !modifiers.HasModifier(TargetModifiers.ForceAttack))
+				// Targeting the terrain is only possible with force-attack modifier or when TargetsTerrain is set
+				if (modifiers.HasModifier(TargetModifiers.ForceMove) || !(modifiers.HasModifier(TargetModifiers.ForceAttack) || ab.Info.TargetsTerrain))
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Traits/Attack/AttackBase.cs]: perhaps move it as the first check and split the 3 check into newlines

**Review by PunkPun — COMMENTED** [review]
LGTM, just a few nits

File: OpenRA.Mods.Common/Traits/Attack/AttackBase.cs (hunk 1 above)
Comment (obrakmann) [reply OpenRA.Mods.Common/Traits/Attack/AttackBase.cs]: Depends on how you setup the conditions, but yes. See new testcase commit.

File: OpenRA.Mods.Common/Traits/Attack/AttackBase.cs (hunk 2 above)
Comment (obrakmann) [reply OpenRA.Mods.Common/Traits/Attack/AttackBase.cs]: I find the one-line version to be more readable, but ok.

File: OpenRA.Mods.Common/Traits/Attack/AttackBase.cs [hunk 3]
```diff
@@ -477,8 +481,10 @@ bool CanTargetLocation(Actor self, CPos location, TargetModifiers modifiers, ref
 
 				IsQueued = modifiers.HasModifier(TargetModifiers.ForceQueue);
 
-				// Targeting the terrain is only possible with force-attack modifier
-				if (modifiers.HasModifier(TargetModifiers.ForceMove) || !modifiers.HasModifier(TargetModifiers.ForceAttack))
+				// Targeting the terrain is only possible with force-attack modifier or when TargetsTerrain is set
+				if (!(ab.Info.TargetsTerrain ||
+					modifiers.HasModifier(TargetModifiers.ForceAttack)) ||
+					modifiers.HasModifier(TargetModifiers.ForceMove))
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Traits/Attack/AttackBase.cs]:
```suggestion
				// Targeting the terrain is only possible with force-attack modifier.
				if (modifiers.HasModifier(TargetModifiers.ForceMove)
					|| (!ab.Info.TargetsTerrain && !modifiers.HasModifier(TargetModifiers.ForceAttack)))
```

I was more thinking like this

File: OpenRA.Mods.Common/Traits/Attack/AttackBase.cs (hunk 1 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Attack/AttackBase.cs]: it's still wrong to keep this comment

File: OpenRA.Mods.Common/Traits/Attack/AttackBase.cs [hunk 4]
```diff
@@ -55,6 +55,10 @@ public abstract class AttackBaseInfo : PausableConditionalTraitInfo
 		[Desc("Tolerance for attack angle. Range [0, 512], 512 covers 360 degrees.")]
 		public readonly WAngle FacingTolerance = new(512);
 
+		[Desc("Show the target cursor even on empty terrain cells when enabled.",
+			"Any traits implementing IMove must be absent or disabled for this to work.")]
+		public readonly bool TargetsTerrain = false;
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Traits/Attack/AttackBase.cs]:
```suggestion
		public readonly bool TargetsTerrain = false;
```

We were discussing on discord that this needs a rename. It's almost sounds as if by having it false the unit won't be able to target terrain at all

Personally I don't have suggestions, sorry

File: OpenRA.Mods.Common/Traits/Attack/AttackBase.cs (hunk 1 above)
Comment (obrakmann) [reply OpenRA.Mods.Common/Traits/Attack/AttackBase.cs]: Oh, I see now. Your original comment wasn't a question but a statement. I don't like resorting to force-move, so this is really meant to be used with conditions, hence the comment. Removed.

File: OpenRA.Mods.Common/Traits/Attack/AttackBase.cs (hunk 4 above)
Comment (obrakmann) [reply OpenRA.Mods.Common/Traits/Attack/AttackBase.cs]: Changed to `TargetTerrainWithoutForceFire`. Any better?

**Review by PunkPun — APPROVED** [review]
LGTM

**Comment by PunkPun** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/4d05cfb2eca5311cd0d2c2c8a2c74fa3f7434add...0cf4aba1c0c2184b5f50c30198bb062a6f1a5dab)

**Pull request merged**